network/benchmark/src/lib.rs

```diff
@@ -295,6 +279,89 @@ pub async fn run_netbench_service(
     }
 }
 
+async fn connection_listener(
+    node_config: NodeConfig,
+    network_client: NetworkClient<NetbenchMessage>,
+    time_service: TimeService,
+    shared: Arc<RwLock<NetbenchSharedState>>,
+    handle: Handle,
+) {
+    let config = node_config.netbench.unwrap();
+    let peers_and_metadata = network_client.get_peers_and_metadata();
+    if let Ok(peers) = peers_and_metadata.get_connected_peers_and_metadata() {
+        info!(
+            "netbench connection_listener got {} initial peers",
+            peers.len()
+        );
+        for peer_network_id in peers.keys() {
+            info!(
+                "netbench connection_listener new for initial {:?}",
+                peer_network_id
+            );
+            if config.enable_direct_send_testing {
+                handle.spawn(direct_sender(
+                    node_config.clone(),
+                    network_client.clone(),
+                    time_service.clone(),
+                    peer_network_id.network_id(),
+                    peer_network_id.peer_id(),
+                    shared.clone(),
+                ));
+            }
+            if config.enable_rpc_testing {
+                handle.spawn(rpc_sender(
+                    node_config.clone(),
+                    network_client.clone(),
+                    time_service.clone(),
+                    peer_network_id.network_id(),
+                    peer_network_id.peer_id(),
+                    shared.clone(),
+                ));
+            }
+        }
+    }
+    let mut connection_notifications = peers_and_metadata.subscribe();
```

Comment:
Hmm... There's a race condition here, e.g., you read the set, a peer connects, and then you subscribe.

Instead of having to manually read the initial set before subscribing, why not have `subscribe()` immediately send a notification with the initial set? This would be cleaner as you'd avoid the race and also not need to manually read?

Reply:
notification doesn't include a copy of the current set.
there's a race condition, but it's a very small window only on startup, and it's in benchmark (test) code, so I'm not too worried about it and I thought it was good-enough.

Reply:
Hmm... leaving a race condition in here makes me feel uncomfortable. Plus, there's no guarantee that it'll remain test-only code. New applications will be added, and the author will likely just copy what you've done, making them susceptible to this issue.

Reply:
Plus, it'll make this code much cleaner. No need to duplicate code like this:
```
        for peer_network_id in peers.keys() {
            info!(
                "netbench connection_listener new for initial {:?}",
                peer_network_id
            );
            if config.enable_direct_send_testing {
                handle.spawn(direct_sender(
                    node_config.clone(),
                    network_client.clone(),
                    time_service.clone(),
                    peer_network_id.network_id(),
                    peer_network_id.peer_id(),
                    shared.clone(),
                ));
            }
            if config.enable_rpc_testing {
                handle.spawn(rpc_sender(
                    node_config.clone(),
                    network_client.clone(),
                    time_service.clone(),
                    peer_network_id.network_id(),
                    peer_network_id.peer_id(),
                    shared.clone(),
                ));
            }
        }
```

Reply:
the race in here is fixed easily enough; `subscribe()` _then_ fetch the initial set of connected peers (and handle any duplicate NewPeer safely).

Reply:
> subscribe() then fetch the initial set of connected peers (and handle any duplicate NewPeer safely).

See my comment below, but we can avoid this race entirely and ensure that notifications are not duplicated. TL;DR all methods that access the peers and metadata object need to grab a lock on. `subscribe()` should be no different. This will avoid race conditions and duplicates, as well as make the calling code simpler (i.e., only a single handler needed to deal with events).